core/services/mongodb-sync/src/main/java/com/mongodb/stitch/core/services/mongodb/sync/internal/CoreSyncImpl.java

```diff
@@ -0,0 +1,194 @@
+package com.mongodb.stitch.core.services.mongodb.sync.internal;
+
+import com.mongodb.MongoNamespace;
+import com.mongodb.stitch.core.services.internal.CoreStitchServiceClient;
+import com.mongodb.stitch.core.services.mongodb.remote.RemoteDeleteResult;
+import com.mongodb.stitch.core.services.mongodb.remote.RemoteInsertOneResult;
+import com.mongodb.stitch.core.services.mongodb.remote.RemoteUpdateResult;
+import com.mongodb.stitch.core.services.mongodb.remote.internal.CoreRemoteFindIterable;
+import com.mongodb.stitch.core.services.mongodb.sync.ChangeEventListener;
+import com.mongodb.stitch.core.services.mongodb.sync.CoreSync;
+import com.mongodb.stitch.core.services.mongodb.sync.ConflictHandler;
+
+import org.bson.BsonDocument;
+import org.bson.BsonValue;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.conversions.Bson;
+
+import java.util.Set;
+
+import javax.annotation.Nullable;
+
+public class CoreSyncImpl<DocumentT> implements CoreSync<DocumentT> {
+    private final DataSynchronizer dataSynchronizer;
+    private final SyncOperations<DocumentT> syncOperations;
+    private final MongoNamespace namespace;
+    private final CodecRegistry codecRegistry;
+    private final Class<DocumentT> documentClass;
+    private final CoreStitchServiceClient service;
+
+    private ConflictHandler<DocumentT> conflictResolver;
+    private ChangeEventListener<DocumentT> changeEventListener;
+
+    public CoreSyncImpl(MongoNamespace namespace,
+                        Class<DocumentT> documentClass,
+                        DataSynchronizer dataSynchronizer,
+                        CoreStitchServiceClient service,
+                        SyncOperations<DocumentT> syncOperations,
+                        CodecRegistry codecRegistry) {
+        this.namespace = namespace;
+        this.documentClass = documentClass;
+        this.dataSynchronizer = dataSynchronizer;
+        this.syncOperations = syncOperations;
+        this.codecRegistry = codecRegistry;
+        this.service = service;
+    }
+
+    @Override
+    public void configure(ConflictHandler<DocumentT> conflictResolver,
+                          ChangeEventListener<DocumentT> changeEventListener) {
+        this.conflictResolver = conflictResolver;
+        this.changeEventListener = changeEventListener;
+    }
+
+    @Override
+    public void syncOne(BsonValue id) {
+        this.dataSynchronizer.syncDocumentFromRemote(
+                this.namespace, id,
+                this.conflictResolver,
+                this.codecRegistry.get(this.documentClass)
```

Comment:
[suggestion] We can probably move this codec registry .get to be a one time thing for the data synchronizer as well during construction?